drs_cli/client.py

```diff
@@ -1,229 +1,424 @@
+            )
+        except (
+            requests.exceptions.ConnectionError,
+            socket.gaierror,
+            urllib3.exceptions.NewConnectionError,
+        ):
+            raise requests.exceptions.ConnectionError(
+                "Could not connect to API endpoint."
+            )
+        if not response.status_code == 200:
+            try:
+                response_val = Error(**response.json())
+            except (
+                json.decoder.JSONDecodeError,
+                pydantic.ValidationError,
+            ):
+                raise InvalidResponseError(
+                    "Response could not be validated against API schema."
+                )
+            logger.warning("Received error response.")
         else:
-            return Error(**response.json())
+            try:
+                response_val = DrsObject(**response.json())
+            except pydantic.ValidationError:
+                raise InvalidResponseError(
+                    "Response could not be validated against API schema."
+                )
+            logger.info(f"Retrieved object: {object_id}")
+        return response_val
 
     def get_access_url(
         self,
         object_id: str,
         access_id: str,
         token: Optional[str] = None,
     ) -> Union[AccessURL, Error]:
-        """Obtain access URL of DRS object.
+        """Retrieve access URL of DRS object.
 
-        Args:
-            object_id: Identifier of DRS object to be retrieved.
-            access_id: Identifier of method giving access to DRS object.
-            token: Bearer token for authentication.
+        Arguments:
+            object_id: Implementation-specific DRS identifier OR hostname-based
+               DRS URI pointing to a given object, cf.
+               https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_hostname_based_drs_uris.
+            access_id: Identifier of one of a DRS object's access methods.
+            token: Bearer token for authentication. Set if required by DRS
+                implementation and if not provided when instatiating client or
+                if expired.
 
         Returns:
             Unmarshalled DRS response as either an instance of `AccessURL`
             in case of a `200` response, or an instance of `Error` for all
             other reponses.
 
         Raises:
-            pydantic.ValidationError: The returned response does not conform
-                to the models defined in the API specification.
+            requests.exceptions.ConnectionError: A connection to the provided
+                DRS instance could not be established.
+            drs_cli.errors.InvalidResponseError: The response could not be
+                validated against the API schema.
         """
-        request_url = f"{self.url}/objects/{object_id}/access/{access_id}"
+        obj_id = self._get_object_id(object_id=object_id)
+        acc_id = quote(string=access_id, safe='')
```

Comment:
I think `quote` is not required here, if the IDs don't contain special characters as per https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_drs_ids.

Also `quote` doc here https://docs.python.org/3/library/urllib.parse.html#urllib.parse.quote says:
> Replace special characters in a string using the %xx escape. Letters, digits, and the characters **'_.-~'** are never quoted.

which are all the characters that follow the DRS specs.

Reply:
I know, but as I said before: DRS IDs _can_ contain other characters, in which case they should be percent-encoded.